### DISCUSSION TITLE
Openshift Integration

### DISCUSSION THREAD

**Discussion by saada** [fluxcd/flux2#531] [Proposals]
As you've probably heard, [Flux is now on OperatorHub](https://operatorhub.io/operator/flux). This means Openshift customers can install and manage Flux resources natively.

We want to allow Openshift users to have an integrated gitops experience when they're bootstrapping Flux without switching too much context at the bootstrap phase.

Today they can install Flux via UI or by applying a subscription CR to their cluster. But using the UI misses out on the Git Repo creation and wiring that `flux bootstrap` does well.

## Proposed CLI

```sh
flux bootstrap --openshift-subscription ...
```

## Behavior

- Automatically apply a [subscription](https://github.com/saada/flux2-openshift/blob/main/test/operator-subscription.yaml)
- When the subscription is applied, Openshift's Operator Lifecycle Manager (OLM) takes care of automatically upgrading Flux for you. You could set upgrades to [manual](https://docs.openshift.com/container-platform/4.5/operators/admin/olm-upgrading-operators.html#olm-approving-pending-upgrade_olm-upgrading-operators) but that requires approvals through the UI which is not a good experience and diverges from the Gitops best practices. I could not find a way to allow Flux to manage itself while still show up as an OLM operator. Instead, we have to rely on OLM to automatically manage the lifecycle of the operator for installs and upgrades.
- Instead of applying the default flux yaml (kustomizations), we only apply the subscription
- Flux cli would also take care of adding the GitRepository and Kustomization objects to the newly created repo during bootstrap step
- Whenever OLM spins up Flux (we can wait for all deployments in flux-system ns), we apply the GitRepository object so it starts syncing, and flux will start managing any yaml resources as well as the OLM subscription object

## Downsides of current design

There doesn't seem to be an easy way to configure the controllers themselves (i.e. custom source-controller flags). Other operators solve this by introducing a higher-level abstraction represented as a CRD. For example a CRD called `Flux`, `kubectl get fluxes`, `kubectl delete flux myflux`. I think you see where I'm going... And then you could have controller specific customizations set as CRD params such as `Flux.spec.helmController.config`.

**Comment by squaremo** [comment]
Thanks Mahmoud, I appreciate you thinking this through and posting here.

Now you have spelt it out, the conflict between OLM wanting to manage the Flux controllers, and managing them with gitops gives me pause, and sounds like it would be difficult to reconcile. In particular, if you committed the subscription to git, but OLM was actively changing its definition in the cluster, that's a recipe for confusion.

Introducing a controller-installer controller ("higher-level abstraction [...] as a CRD") is more complication than I think we'd want here. Perhaps my imagination is lacking, but I can't see a useful purpose, for example, to having arbitrary numbers of instances of gotk controllers running in a cluster. It doesn't make sense in the way that e.g., running instances of Prometheus might make sense.

For the reasons above I would be reluctant to introduce this as a feature for `flux` specifically, but that needn't be a roadblock to a solution being developed -- that might be a commercial opportunity, even.

**Comment by kingdonb** [comment]
> For example a CRD called `Flux`, `kubectl get fluxes`, `kubectl delete flux myflux`

While I am in love with the idea that OpenShift's OLM can manage multiple concurrent installed versions of an operator by setting up subscriptions and tying them to different namespaces, I've yet to see it demonstrated that this can actually work with Flux, which we understand as global for the cluster, and not designed to have multiple concurrent instances running on a single cluster.

And there is some room to argue here that this is still a good idea for consistent approach to configuring Flux. But what we'd be seeing here in most cases is a single instance of the `Flux` CRD on any given cluster.

I think we'd say in most cases that we don't want Flux to be configured, as it should behave consistently across all installations and we shouldn't expect to see strange behavior that differs from one installation to another; in the ideal that's what I prefer, Flux has a clearly defined one job and it always does it the same way, where any need for special configuration is handled on a per-resource level

(Like a GitRepository which is known to be a bit large and slow to clone having a raised value of `spec.timeout` to avoid raising the occasional error.)

But still we find in spite of this ideal, that there are some global configuration options and it's maybe a bit counter-intuitive that you must patch your Flux install in order to access them safely.

Thanks again @saada for contributing this Operator Hub version of Flux, we'd love to revisit this with you any time and show how we're maintaining the project that you built us. This is making Flux more accessible to lots of different people. I'm still really interested to know more about how operator lifecycle manager subscriptions work and what kind of uses they have found among operator projects.

**Reply by kingdonb** [reply]
We have documented some around [how to use the kustomize overlay in `flux-system`](https://github.com/fluxcd/website/blob/main/content/en/docs/installation.md/#customize-flux-manifests) to override Flux defaults and set up specialized or non-standard configuration in the Flux way. When users find this instruction in the Flux guide will they balk at required configuration directives because they haven't yet acquired the skills that are needed to customize the manifests? (Or if they came in via Operator Hub, they will not have a `gotk-components.yaml` in their repo and may be utterly lost about how to approach setting these customized options on any given controller. I myself am not sure how I would accomplish this on OpenShift, nor have any idea of how to guide any user that came with this question.)

There is an argument that patching manifests is a core feature of Flux, and it's imperative we coach our users to gain this skill as readily as possible, but a counter-argument is this prescription is out of order, and new users of Flux need to configure it when they are just starting, still have yet to finely hone these skills, (or simply can't do it because OLM does not have any place for it!)

These are the type of options I'd expect to find in the form of a `values.yaml` if Flux were being delivered in the form of a Helm chart, but since we aren't using that method, we may have to be conscious that there [isn't really a place for those configuration ideas to live](https://github.com/fluxcd/flux2/discussions/304) in the sense that GitOps prescribes; I would expect the intents to live in a file, not to be decided by some options that we may forget to pass in faithfully again at bootstrap time when we are upgrading from one version to next.

I'm happy that we have Flux as a power tool and I think it's fair that we are basically "cutting with/not against the grain" to accept that doing effective Flux configuration requires patching YAML manifests and Flux users will need to acquire this skill rapidly in order to make the most out of Flux, so we better help them get on with it, and it's actually beneficial that there is this hurdle configuring Flux. But that's also likely not going to be the experience that all Flux users were expecting, (especially not those folks who installed by clicking in the Operator Hub panel of their OpenShift operator consoles.)

I agree that:
> controller-installer controller ("higher-level abstraction [...] as a CRD") is more complication than I think we'd want

but it is worth considering how much additional other complexity are we accepting leaving on the table in return for not implementing such a thing?

https://github.com/fluxcd/website/blob/main/content/en/docs/installation.md/#customize-flux-manifests

[Using Native GCP GCR Auto-Login](https://github.com/fluxcd/website/pull/702/files?short_path=a994294#diff-a994294fcec130efcd91a396eefcf70eaf036cee5aec03bee28d44f6a5b5e5ae)

I personally can understand these guides fine and don't have any problem with these instructions, but I'm not sure I'm entirely representative of our target audience. (I'm saying these are well above-average difficulty steps, we're going to need and accumulate more like this in our guides as time goes on unless something changes, and while I am someone who drank plenty of kool-aid so I didn't balk, I'm not sure that's definitive. Maybe I am totally wrong about this, and our users can all totally handle it?)

The same problem rears its head again in a different venue, customizing Flux takes yet another form for Flux's terraform provider: https://github.com/fluxcd/terraform-provider-flux/issues/231
